# 1. Import the meta data of items

In [1]:
import pickle
f = open('pickle_data/all_items_info.p', 'rb')
items_info = pickle.load(f)
f.close()

In [2]:
len(items_info)

8528

In [3]:
items_info[0]

{'color': 'White',
 'image': {'item': 'http://lp.hm.com/hmprod?set=key[source],value[/model/2017/E00 0478110 001 77 1688.jpg]&set=key[rotate],value[]&set=key[width],value[]&set=key[height],value[]&set=key[x],value[]&set=key[y],value[]&set=key[type],value[STILL_LIFE_FRONT]&set=key[hmver],value[2]&call=url[file:/product/large]',
  'model': 'http://lp.hm.com/hmprod?set=key[source],value[/environment/2017/8GZ_0158_011R.jpg]&set=key[rotate],value[0]&set=key[width],value[3927]&set=key[height],value[4591]&set=key[x],value[0]&set=key[y],value[-1]&set=key[type],value[FASHION_FRONT]&set=key[hmver],value[0]&call=url[file:/product/large]'},
 'metricCategoryID': 'LADIES_TOPS_SHORT_SLEEVED_W112',
 'name': 'Short-sleeved Top',
 'serial': '65274-B',
 'url': 'http://www.hm.com/us/product/65274?article=65274-B',
 'who': 'Women'}

# 2. Digging all kinds of categories

In [12]:
def get_categories(item_info):
    metricCategoryID = item_info['metricCategoryID']
    # Delete the first and last strings
    categories = metricCategoryID.split('_')[1:-1]
    return categories

In [13]:
for item_info in items_info[0:10]:
    print get_categories(item_info)

['TOPS', 'SHORT', 'SLEEVED']
['TOPS', 'SHORT', 'SLEEVED']
['TOPS', 'VESTS', 'SLEEVELESS']
['TOPS', 'LONG', 'SLEEVED']
['TOPS', 'VESTS', 'SLEEVELESS']
['DIVIDED', 'BASICS', 'TOPS']
['TOPS', 'SHORT', 'SLEEVED']
['TOPS', 'SHORT', 'SLEEVED']
['TOPS', 'SHORT', 'SLEEVED']
['CASUAL', 'SHIRTS', 'BLOUSES']


In [14]:
set_category = set()
set_category_first = set()
set_category_last = set()
for item_info in items_info:
    categories = get_categories(item_info)
    set_category_first.add(categories[0])
    set_category_last.add(categories[-1])
    for category in categories:
        set_category.add(category)
print "set_category:\n", set_category, '\n'
print "set_category_first:\n", set_category_first, '\n'
print "set_category_last:\n", set_category_last, '\n'

set_category:
set(['EYEBROWS', 'EYESHADOW', 'TOWELS', 'BEACHWEAR', 'PERFUME', 'COATS', 'FRAGRANCE', 'CARDIGANS', 'EVERYDAY', 'SPORTSWEAR', 'SLIPS', 'GLOVES', 'CONCEALER', 'SCRUB', 'DENIMSHORTS', 'HANDBAGS', 'MOBILEACC', 'LONG', 'PUMPS', 'DRESSING', 'EARRINGS', 'SHAPING', 'BALLERINAS', 'BRUSHES', 'LASHES', 'LIMITED', 'PARTYDRESSES', 'LIPS', 'MASCARA', 'TUNICS', 'SWEATSHIRT', 'PRIMER', 'LIPLINER', 'POWDER', 'REMOVERS', 'PADDED', 'LIPSTICK', 'CHINOS', 'STRAIGHT', 'SOCKS', 'NAILPOLISH', 'REGULAR', 'MATERNITY', 'JACKETS', 'BEDROOM', 'MODERN-CLASSICS', 'BOOTCUT', 'SHORT', 'BODYLOTIONS', 'HATS', 'TOPS', 'BRIEFS', 'BATH', 'BOOTS', 'SPONGES', 'HM', 'TIES', 'BATHROOM', 'PQ', 'MULTIPACK', 'BLOUSES', 'FACEMASKS', 'SWEATSHIRTS', 'ROBES', 'PURSES', 'LINGERIE', 'CORSETS', 'PUSHUP', 'SNEAKERS', 'APPLICATORS', 'ACC', 'JUMPERS', 'EYES', 'VESTS', 'BELTS', 'TURTLENECKS', 'KIMONOS', 'SCARVES', 'MODERNCLASSICS', 'BRONZER', 'JUMPSUITS', 'SWEATPANTS', 'HIPSTERS', 'SLEEVELESS', 'JEANS', 'WATCHES', 'CROPTOPS', 

### Find inclusive or independent categories

In [15]:
def get_supersets(category):
    is_first_iter = True
    supersets = set()
    for item_info in items_info:
        categories = get_categories(item_info)
        if category in categories:
            if is_first_iter:
                supersets = set(categories)
                is_first_iter = False
            else:
                supersets = supersets.intersection(categories)
    return supersets

In [16]:
def get_proper_supersets(category):
    is_first_iter = True
    supersets = set()
    for item_info in items_info:
        categories = get_categories(item_info)
        if category in categories:
            if is_first_iter:
                supersets = set(categories)
                is_first_iter = False
            else:
                supersets = supersets.intersection(categories)
    return supersets.difference(set({category}))

In [17]:
get_supersets('LEGGINGS')

{'LEGGINGS'}

In [18]:
get_proper_supersets('LEGGINGS')

set()

In [19]:
# print the superset categories
superset_categories = []
for category in set_category:
    supersets = get_supersets(category)
    if len(supersets) == 1:
        print "category:", category
        superset_categories.append(category)

category: SPORTSWEAR
category: LONG
category: TUNICS
category: SOCKS
category: JACKETS
category: MODERN-CLASSICS
category: SHORT
category: TOPS
category: BATH
category: HM
category: BATHROOM
category: PQ
category: LINGERIE
category: ACC
category: VESTS
category: MODERNCLASSICS
category: JEANS
category: BOTTOMS
category: SKIRTS
category: SUITS
category: HAIR
category: LEGGINGS
category: ACCESSORIES
category: NIGHTWEAR
category: SHORTS
category: DIVIDED
category: SLIM
category: BASICS
category: SWIMWEAR
category: BLAZERS
category: TSHIRT
category: SHIRTS
category: CASUAL
category: JUMPSUIT
category: TIGHTS
category: UNDERWEAR
category: BEAUTY
category: DRESSED
category: TROUSERS
category: DRESSES
category: SLEEVED
category: HOODIES
category: SHOES
category: MIDI
category: LOOSE
category: PARTYWEAR


In [20]:
print superset_categories

['SPORTSWEAR', 'LONG', 'TUNICS', 'SOCKS', 'JACKETS', 'MODERN-CLASSICS', 'SHORT', 'TOPS', 'BATH', 'HM', 'BATHROOM', 'PQ', 'LINGERIE', 'ACC', 'VESTS', 'MODERNCLASSICS', 'JEANS', 'BOTTOMS', 'SKIRTS', 'SUITS', 'HAIR', 'LEGGINGS', 'ACCESSORIES', 'NIGHTWEAR', 'SHORTS', 'DIVIDED', 'SLIM', 'BASICS', 'SWIMWEAR', 'BLAZERS', 'TSHIRT', 'SHIRTS', 'CASUAL', 'JUMPSUIT', 'TIGHTS', 'UNDERWEAR', 'BEAUTY', 'DRESSED', 'TROUSERS', 'DRESSES', 'SLEEVED', 'HOODIES', 'SHOES', 'MIDI', 'LOOSE', 'PARTYWEAR']


### Analzing categories

In [32]:
import random
def print_has(category, max_items):
    count = 0
    while True:
        item_info = random.choice(items_info)
        categories = get_categories(item_info)
        if category in categories:
            print categories
            print item_info['url']
            count = count + 1
            if count == max_items:
                return

In [35]:
print_has('TUNICS', 10)

['SHIRTS', 'BLOUSES', 'TUNICS']
http://www.hm.com/us/product/68372?article=68372-D
['SHIRTS', 'BLOUSES', 'TUNICS']
http://www.hm.com/us/product/67416?article=67416-C
['HM', 'PLUS', 'TUNICS']
http://www.hm.com/us/product/60300?article=60300-A
['HM', 'PLUS', 'TUNICS']
http://www.hm.com/us/product/60300?article=60300-A
['SHIRTS', 'BLOUSES', 'TUNICS']
http://www.hm.com/us/product/67416?article=67416-C
['SHIRTS', 'BLOUSES', 'TUNICS']
http://www.hm.com/us/product/61117?article=61117-A
['HM', 'PLUS', 'TUNICS']
http://www.hm.com/us/product/60318?article=60318-A
['SHIRTS', 'BLOUSES', 'TUNICS']
http://www.hm.com/us/product/61053?article=61053-A
['SHIRTS', 'BLOUSES', 'TUNICS']
http://www.hm.com/us/product/67282?article=67282-B
['HM', 'PLUS', 'TUNICS']
http://www.hm.com/us/product/64818?article=64818-A


* 'TUNICS' is a one-piece style.

In [38]:
print_has('SHORT', 10)

['TSHIRT', 'SHORT', 'SLEEVED']
http://www.hm.com/us/product/39694?article=39694-E
['TSHIRT', 'SHORT', 'SLEEVED']
http://www.hm.com/us/product/46072?article=46072-C
['TOPS', 'SHORT', 'SLEEVED']
http://www.hm.com/us/product/68429?article=68429-A
['TOPS', 'SHORT', 'SLEEVED']
http://www.hm.com/us/product/70738?article=70738-A
['TSHIRT', 'SHORT', 'SLEEVED']
http://www.hm.com/us/product/56721?article=56721-A
['TOPS', 'SHORT', 'SLEEVED']
http://www.hm.com/us/product/62358?article=62358-B
['SKIRTS', 'SHORT']
http://www.hm.com/us/product/61713?article=61713-A
['DRESSES', 'JUMPSUITS', 'SHORT']
http://www.hm.com/us/product/69846?article=69846-B
['TOPS', 'SHORT', 'SLEEVED']
http://www.hm.com/us/product/69128?article=69128-A
['TOPS', 'SHORT', 'SLEEVED']
http://www.hm.com/us/product/51149?article=51149-G


In [39]:
print_has('SHORTS', 10)

['EVERYDAY', 'FASHION', 'SHORTS']
http://www.hm.com/us/product/68195?article=68195-D
['DIVIDED', 'TROUSERS', 'SHORTS']
http://www.hm.com/us/product/69010?article=69010-A
['DIVIDED', 'TROUSERS', 'SHORTS']
http://www.hm.com/us/product/69010?article=69010-E
['SHORTS', 'DENIMSHORTS']
http://www.hm.com/us/product/64586?article=64586-D
['SHORTS', 'OTHERS']
http://www.hm.com/us/product/65973?article=65973-A
['SHORTS', 'DENIMSHORTS']
http://www.hm.com/us/product/29840?article=29840-E
['HM', 'MAN', 'SHORTS']
http://www.hm.com/us/product/57540?article=57540-A
['CASUAL', 'SHORTS']
http://www.hm.com/us/product/39670?article=39670-T
['DIVIDED', 'TROUSERS', 'SHORTS']
http://www.hm.com/us/product/68328?article=68328-A
['SHORTS', 'DENIMSHORTS']
http://www.hm.com/us/product/64153?article=64153-C


* 'SHORT' is a style, and 'SHORTS' is a bottom.

In [41]:
print_has('PQ',10)

['SHOES', 'PQ']
http://www.hm.com/us/product/61758?article=61758-C
['SHOES', 'PQ']
http://www.hm.com/us/product/61758?article=61758-A
['ACCESSORIES', 'PQ']
http://www.hm.com/us/product/70417?article=70417-A
['SHOES', 'PQ']
http://www.hm.com/us/product/54574?article=54574-B
['SHOES', 'PQ']
http://www.hm.com/us/product/61758?article=61758-C
['SHOES', 'PQ']
http://www.hm.com/us/product/62454?article=62454-B
['SHOES', 'PQ']
http://www.hm.com/us/product/57950?article=57950-A
['ACCESSORIES', 'PQ']
http://www.hm.com/us/product/66243?article=66243-A
['SHOES', 'PQ']
http://www.hm.com/us/product/61936?article=61936-A
['SHOES', 'PQ']
http://www.hm.com/us/product/62454?article=62454-A


* 'PQ' means 'Premium Quality'.

In [43]:
print_has('VESTS',10)

['TOPS', 'VESTS', 'SLEEVELESS']
http://www.hm.com/us/product/68924?article=68924-A
['TOPS', 'VESTS', 'SLEEVELESS']
http://www.hm.com/us/product/67562?article=67562-F
['HM', 'MAN', 'TSHIRTS', 'VESTS']
http://www.hm.com/us/product/68533?article=68533-C
['HM', 'MAN', 'TSHIRTS', 'VESTS']
http://www.hm.com/us/product/64832?article=64832-C
['HM', 'MAN', 'TSHIRTS', 'VESTS']
http://www.hm.com/us/product/53945?article=53945-A
['TOPS', 'VESTS', 'SLEEVELESS']
http://www.hm.com/us/product/69249?article=69249-A
['TOPS', 'VESTS', 'SLEEVELESS']
http://www.hm.com/us/product/68585?article=68585-I
['CASUAL', 'TSHIRTS', 'VESTS']
http://www.hm.com/us/product/64842?article=64842-A
['TSHIRT', 'VESTS']
http://www.hm.com/us/product/67985?article=67985-B
['TOPS', 'VESTS', 'SLEEVELESS']
http://www.hm.com/us/product/69124?article=69124-D


* 'VESTS' is either a top or a T-shirt.

In [45]:
print_has('SUITS',10)

['BLAZERS', 'SUITS', 'BLAZERS']
http://www.hm.com/us/product/41772?article=41772-A
['HM', 'MAN', 'SUITS', 'TROUSERS']
http://www.hm.com/us/product/65248?article=65248-B
['BLAZERS', 'SUITS', 'BLAZERS']
http://www.hm.com/us/product/41772?article=41772-A
['BLAZERS', 'SUITS', 'BLAZERS']
http://www.hm.com/us/product/60546?article=60546-A
['BLAZERS', 'SUITS', 'BLAZERS']
http://www.hm.com/us/product/56395?article=56395-A
['BLAZERS', 'SUITS', 'BLAZERS']
http://www.hm.com/us/product/59700?article=59700-A
['HM', 'MAN', 'SUITS', 'JACKETS']
http://www.hm.com/us/product/60224?article=60224-B
['HM', 'MAN', 'SUITS', 'TROUSERS']
http://www.hm.com/us/product/46504?article=46504-B
['HM', 'MAN', 'SUITS', 'JACKETS']
http://www.hm.com/us/product/65247?article=65247-A
['HM', 'MAN', 'SUITS', 'TROUSERS']
http://www.hm.com/us/product/52740?article=52740-A


* 'SUITS' means a style.

In [46]:
print_has('HAIR',10)

['ACCESSORIES', 'HAIR']
http://www.hm.com/us/product/82432?article=82432-L
['ACCESSORIES', 'HAIR']
http://www.hm.com/us/product/82432?article=82432-I
['ACCESSORIES', 'HAIR']
http://www.hm.com/us/product/58185?article=58185-A
['ACCESSORIES', 'HAIR']
http://www.hm.com/us/product/82450?article=82450-P
['ACCESSORIES', 'HAIR']
http://www.hm.com/us/product/28971?article=28971-G
['ACCESSORIES', 'HAIR']
http://www.hm.com/us/product/82432?article=82432-K
['BEAUTY', 'HAIR']
http://www.hm.com/us/product/81574?article=81574-A
['ACCESSORIES', 'HAIR']
http://www.hm.com/us/product/85546?article=85546-T
['ACCESSORIES', 'HAIR']
http://www.hm.com/us/product/67624?article=67624-F
['BEAUTY', 'HAIR']
http://www.hm.com/us/product/81566?article=81566-A


* 'HAIR' is a category that is related to hair.

In [47]:
print_has('DIVIDED',10)

['DIVIDED', 'JEANS', 'SKINNY', 'SKINNY', 'REGULAR']
http://www.hm.com/us/product/48763?article=48763-R
['DIVIDED', 'DRESSES', 'JUMPSUITS']
http://www.hm.com/us/product/59442?article=59442-A
['DIVIDED', 'CARDIGANS', 'JUMPERS', 'JUMPERS']
http://www.hm.com/us/product/55632?article=55632-B
['DIVIDED', 'DRESSES', 'JUMPSUITS']
http://www.hm.com/us/product/64620?article=64620-E
['DIVIDED', 'SHIRTS']
http://www.hm.com/us/product/65172?article=65172-C
['DIVIDED', 'DRESSES', 'JUMPSUITS']
http://www.hm.com/us/product/66759?article=66759-A
['DIVIDED', 'BASICS', 'HOODIES', 'SWEATSHIRTS']
http://www.hm.com/us/product/62333?article=62333-F
['DIVIDED', 'BASICS', 'TOPS']
http://www.hm.com/us/product/34980?article=34980-B
['DIVIDED', 'BASICS', 'TOPS']
http://www.hm.com/us/product/87648?article=87648-T
['DIVIDED', 'DRESSES', 'JUMPSUITS']
http://www.hm.com/us/product/02632?article=02632-K


* 'DIVIDED' is not a category only included in tops or bottoms.

In [49]:
print_has('TIGHTS',10)

['LINGERIE', 'SOCKS', 'TIGHTS', 'SOCKS']
http://www.hm.com/us/product/54598?article=54598-C
['MATERNITY', 'WEAR', 'TIGHTS']
http://www.hm.com/us/product/47408?article=47408-A
['LINGERIE', 'SOCKS', 'TIGHTS', 'SOCKS']
http://www.hm.com/us/product/34243?article=34243-A
['LINGERIE', 'SOCKS', 'TIGHTS', 'TIGHTS']
http://www.hm.com/us/product/63764?article=63764-A
['LINGERIE', 'SOCKS', 'TIGHTS', 'SOCKS']
http://www.hm.com/us/product/66887?article=66887-C
['LINGERIE', 'SOCKS', 'TIGHTS', 'SOCKS']
http://www.hm.com/us/product/43059?article=43059-A
['MATERNITY', 'WEAR', 'TIGHTS']
http://www.hm.com/us/product/46249?article=46249-A
['LINGERIE', 'SOCKS', 'TIGHTS', 'SOCKS']
http://www.hm.com/us/product/54598?article=54598-C
['LINGERIE', 'SOCKS', 'TIGHTS', 'SOCKS']
http://www.hm.com/us/product/69995?article=69995-A
['MATERNITY', 'WEAR', 'TIGHTS']
http://www.hm.com/us/product/41415?article=41415-A


* 'TIGHTS' means a tightly slim style. ('tights' = '타이즈')

In [50]:
print_has('DRESSED',10)

['SHIRTS', 'DRESSED']
http://www.hm.com/us/product/62022?article=62022-A
['SHIRTS', 'DRESSED']
http://www.hm.com/us/product/63611?article=63611-H
['TROUSERS', 'DRESSED']
http://www.hm.com/us/product/62543?article=62543-A
['TROUSERS', 'DRESSED']
http://www.hm.com/us/product/34933?article=34933-A
['TROUSERS', 'DRESSED']
http://www.hm.com/us/product/60693?article=60693-A
['SHIRTS', 'DRESSED']
http://www.hm.com/us/product/64765?article=64765-B
['SHIRTS', 'DRESSED']
http://www.hm.com/us/product/48995?article=48995-P
['SHIRTS', 'DRESSED']
http://www.hm.com/us/product/65202?article=65202-B
['SHIRTS', 'DRESSED']
http://www.hm.com/us/product/62022?article=62022-A
['TROUSERS', 'DRESSED']
http://www.hm.com/us/product/58547?article=58547-A


* 'DRESSED' means a formal style.

In [51]:
print_has('DRESSES',20)

['DRESSES', 'JUMPSUITS', 'MIDI']
http://www.hm.com/us/product/59908?article=59908-A
['DIVIDED', 'DRESSES', 'JUMPSUITS']
http://www.hm.com/us/product/57528?article=57528-F
['DIVIDED', 'DRESSES', 'JUMPSUITS']
http://www.hm.com/us/product/57610?article=57610-D
['MATERNITY', 'WEAR', 'DRESSES']
http://www.hm.com/us/product/44962?article=44962-C
['DRESSES', 'JUMPSUITS', 'SHORT']
http://www.hm.com/us/product/69818?article=69818-B
['DIVIDED', 'DRESSES', 'JUMPSUITS']
http://www.hm.com/us/product/68300?article=68300-A
['DRESSES', 'JUMPSUITS', 'LONG']
http://www.hm.com/us/product/69517?article=69517-A
['EVERYDAY', 'FASHION', 'DRESSES', 'JUMPSUITS']
http://www.hm.com/us/product/57557?article=57557-C
['DIVIDED', 'DRESSES', 'JUMPSUITS']
http://www.hm.com/us/product/62263?article=62263-C
['DIVIDED', 'DRESSES', 'JUMPSUITS']
http://www.hm.com/us/product/43326?article=43326-A
['HM', 'PLUS', 'DRESSES']
http://www.hm.com/us/product/67699?article=67699-A
['DRESSES', 'JUMPSUITS', 'LONG']
http://www.hm.com/u

In [52]:
print_has('SKIRTS',20)

['SKIRTS', 'SHORT']
http://www.hm.com/us/product/67166?article=67166-A
['MODERNCLASSICS', 'SKIRTS']
http://www.hm.com/us/product/65336?article=65336-A
['SKIRTS', 'SHORT']
http://www.hm.com/us/product/62725?article=62725-D
['BASICS', 'DRESSES', 'SKIRTS']
http://www.hm.com/us/product/66300?article=66300-A
['BASICS', 'DRESSES', 'SKIRTS']
http://www.hm.com/us/product/32266?article=32266-X
['SKIRTS', 'LONG']
http://www.hm.com/us/product/64524?article=64524-C
['SKIRTS', 'SHORT']
http://www.hm.com/us/product/19634?article=19634-A
['DIVIDED', 'BASICS', 'DRESSES', 'SKIRTS']
http://www.hm.com/us/product/65509?article=65509-C
['SKIRTS', 'SHORT']
http://www.hm.com/us/product/23376?article=23376-A
['BASICS', 'DRESSES', 'SKIRTS']
http://www.hm.com/us/product/70338?article=70338-B
['SKIRTS', 'MIDI']
http://www.hm.com/us/product/58367?article=58367-A
['DIVIDED', 'BASICS', 'DRESSES', 'SKIRTS']
http://www.hm.com/us/product/70827?article=70827-A
['BASICS', 'DRESSES', 'SKIRTS']
http://www.hm.com/us/produc

* 'SKIRTS' and 'DRESSES' are sometimes overlapped.
* If they are overlapped, we can categorize the item by its name.
* At the end of the name of an item, __'skirt'__ or __'dress'__ is used instead of 'skirts' or 'dresses'.

In [54]:
print_has('MIDI',10)

['DRESSES', 'JUMPSUITS', 'MIDI']
http://www.hm.com/us/product/68281?article=68281-C
['SKIRTS', 'MIDI']
http://www.hm.com/us/product/58367?article=58367-A
['SKIRTS', 'MIDI']
http://www.hm.com/us/product/55496?article=55496-B
['SKIRTS', 'MIDI']
http://www.hm.com/us/product/62489?article=62489-B
['DRESSES', 'JUMPSUITS', 'MIDI']
http://www.hm.com/us/product/67183?article=67183-C
['DRESSES', 'JUMPSUITS', 'MIDI']
http://www.hm.com/us/product/67183?article=67183-C
['DRESSES', 'JUMPSUITS', 'MIDI']
http://www.hm.com/us/product/56814?article=56814-D
['DRESSES', 'JUMPSUITS', 'MIDI']
http://www.hm.com/us/product/69017?article=69017-B
['DRESSES', 'JUMPSUITS', 'MIDI']
http://www.hm.com/us/product/71905?article=71905-B
['SKIRTS', 'MIDI']
http://www.hm.com/us/product/61664?article=61664-B


* 'MIDI' is something related to skirts. It presumably means a style.

These categories are inclusive categories.

The following categories are categories that can be classified one of the three: tops, bottoms, and one-pieces(top-bottom together).

['ACC', 'SPORTSWEAR', 'LONG', __'TUNICS'__, 'LINGERIE', __'JACKETS'__, 'MODERN-CLASSICS', 'SHORT', __'TOPS'__, 'BATH', 'SOCKS', 'HM', 'BATHROOM', 'PQ', __'VESTS'__, 'MODERNCLASSICS', __'JEANS'__, __'BOTTOMS'__, __'SKIRTS'__, 'SUITS', 'HAIR', __'LEGGINGS'__, 'ACCESSORIES', 'NIGHTWEAR', __'SHORTS'__, 'DIVIDED', 'SLIM', 'BASICS', 'SWIMWEAR', __'BLAZERS'__, __'TSHIRT'__, __'SHIRTS'__, 'CASUAL', __'JUMPSUIT'__, 'TIGHTS', 'UNDERWEAR', 'BEAUTY', 'DRESSED', __'TROUSERS'__, __'DRESSES'__, 'SLEEVED', __'HOODIES'__, 'SHOES', 'MIDI', 'LOOSE', 'PARTYWEAR']

#### Tops
* ['JACKETS', 'TOPS', 'VESTS', 'BLAZERS', 'TSHIRT', 'SHIRTS', 'HOODIES']

#### Bottoms
* ['JEANS', 'BOTTOMS', 'SKIRTS', 'LEGGINGS', 'SHORTS', 'TROUSERS']

#### Onepieces
* ['TUNICS', 'JUMPSUIT', 'DRESSES']

In [55]:
dict_tbo_categories = dict()

In [56]:
dict_tbo_categories['top'] = ['JACKETS', 'TOPS', 'VESTS', 'BLAZERS', 'TSHIRT', 'SHIRTS', 'HOODIES']

In [57]:
dict_tbo_categories['bottom'] = ['JEANS', 'BOTTOMS', 'SKIRTS', 'LEGGINGS', 'SHORTS', 'TROUSERS']

In [58]:
dict_tbo_categories['onepiece'] = ['TUNICS', 'JUMPSUIT', 'DRESSES']

In [59]:
print dict_tbo_categories

{'top': ['JACKETS', 'TOPS', 'VESTS', 'BLAZERS', 'TSHIRT', 'SHIRTS', 'HOODIES'], 'onepiece': ['TUNICS', 'JUMPSUIT', 'DRESSES'], 'bottom': ['JEANS', 'BOTTOMS', 'SKIRTS', 'LEGGINGS', 'SHORTS', 'TROUSERS']}


### Serialization

In [60]:
import pickle
f = open('pickle_data/dict_tbo_categories.p', 'wb')
pickle.dump(dict_tbo_categories, f)
f.close()

# 3. Are tops, bottoms, and onepieces mutually exclusive?

In [62]:
from interface import *

In [63]:
dict_tbo_categories = import_tbo_categories()

In [64]:
dict_tbo_categories['top']

['JACKETS', 'TOPS', 'VESTS', 'BLAZERS', 'TSHIRT', 'SHIRTS', 'HOODIES']

In [65]:
dict_tbo_categories['bottom']

['JEANS', 'BOTTOMS', 'SKIRTS', 'LEGGINGS', 'SHORTS', 'TROUSERS']

In [66]:
dict_tbo_categories['onepiece']

['TUNICS', 'JUMPSUIT', 'DRESSES']

In [68]:
pf = open('pickle_data/all_items_info.p', 'rb')
items_info = pickle.load(pf)
pf.close()

In [69]:
def get_categories(item_info):
    metricCategoryID = item_info['metricCategoryID']
    # Delete the first and last strings
    categories = metricCategoryID.split('_')[1:-1]
    return categories

In [70]:
get_categories(items_info[0])

['TOPS', 'SHORT', 'SLEEVED']

In [71]:
def print_categories_with_tbo_num(tbo_num_to_print):
    count = 0
    for item_info in items_info:
        item_info['tbo'] = set()
        categories = get_categories(item_info)
        for category in categories:
            if category in dict_tbo_categories['top']:
                item_info['tbo'].add('top')
            if category in dict_tbo_categories['bottom']:
                item_info['tbo'].add('bottom')
            if category in dict_tbo_categories['onepiece']:
                item_info['tbo'].add('onepiece')
        if len(item_info['tbo']) == tbo_num_to_print:
            print "TBO       :", item_info['tbo']
            print "categories:", categories
            print "name:", item_info['name']
            print "URL       :", item_info['url']
            print '\n'
            count = count + 1
    print "Count:", count 

In [72]:
print_categories_with_tbo_num(2)

TBO       : set(['top', 'onepiece'])
categories: ['SHIRTS', 'BLOUSES', 'TUNICS']
name: V-neck Tunic
URL       : http://www.hm.com/us/product/68372?article=68372-A


TBO       : set(['onepiece', 'bottom'])
categories: ['DIVIDED', 'BASICS', 'DRESSES', 'SKIRTS']
name: Maxi Dress
URL       : http://www.hm.com/us/product/27458?article=27458-N


TBO       : set(['onepiece', 'bottom'])
categories: ['DIVIDED', 'BASICS', 'DRESSES', 'SKIRTS']
name: Short Jersey Dress
URL       : http://www.hm.com/us/product/70827?article=70827-A


TBO       : set(['onepiece', 'bottom'])
categories: ['DIVIDED', 'BASICS', 'DRESSES', 'SKIRTS']
name: Short Jersey Dress
URL       : http://www.hm.com/us/product/70827?article=70827-A


TBO       : set(['onepiece', 'bottom'])
categories: ['BASICS', 'DRESSES', 'SKIRTS']
name: Short-sleeved Jersey Dress
URL       : http://www.hm.com/us/product/32266?article=32266-F


TBO       : set(['onepiece', 'bottom'])
categories: ['DIVIDED', 'BASICS', 'DRESSES', 'SKIRTS']
name: T-shi

* The 94 items are more than one TBO categories.

In [73]:
print_categories_with_tbo_num(3)

Count: 0


###  Only for counting


In [74]:
def count_categories_with_tbo_num(tbo_num_to_print):
    count = 0
    for item_info in items_info:
        item_info['tbo'] = set()
        categories = get_categories(item_info)
        for category in categories:
            if category in dict_tbo_categories['top']:
                item_info['tbo'].add('top')
            if category in dict_tbo_categories['bottom']:
                item_info['tbo'].add('bottom')
            if category in dict_tbo_categories['onepiece']:
                item_info['tbo'].add('onepiece')
        if len(item_info['tbo']) == tbo_num_to_print:
            count = count + 1
    print "Count:", count 

In [79]:
count_categories_with_tbo_num(0)

Count: 3326


In [80]:
count_categories_with_tbo_num(1)

Count: 5108


In [81]:
count_categories_with_tbo_num(2)

Count: 94


In [82]:
count_categories_with_tbo_num(3)

Count: 0


#### Reult: The number of items that has particular numbers of TBO categories
* 0 TBO categories: 3326 (These are like accesories, shoes, hats, etc..)
* 1 TBO category: 5108
* 2 TBO categories: 94
* 3 TBO categories: 0

### Preventing from having 2 TBO categories

In [83]:
print_categories_with_tbo_num(2)

TBO       : set(['top', 'onepiece'])
categories: ['SHIRTS', 'BLOUSES', 'TUNICS']
name: V-neck Tunic
URL       : http://www.hm.com/us/product/68372?article=68372-A


TBO       : set(['onepiece', 'bottom'])
categories: ['DIVIDED', 'BASICS', 'DRESSES', 'SKIRTS']
name: Maxi Dress
URL       : http://www.hm.com/us/product/27458?article=27458-N


TBO       : set(['onepiece', 'bottom'])
categories: ['DIVIDED', 'BASICS', 'DRESSES', 'SKIRTS']
name: Short Jersey Dress
URL       : http://www.hm.com/us/product/70827?article=70827-A


TBO       : set(['onepiece', 'bottom'])
categories: ['DIVIDED', 'BASICS', 'DRESSES', 'SKIRTS']
name: Short Jersey Dress
URL       : http://www.hm.com/us/product/70827?article=70827-A


TBO       : set(['onepiece', 'bottom'])
categories: ['BASICS', 'DRESSES', 'SKIRTS']
name: Short-sleeved Jersey Dress
URL       : http://www.hm.com/us/product/32266?article=32266-F


TBO       : set(['onepiece', 'bottom'])
categories: ['DIVIDED', 'BASICS', 'DRESSES', 'SKIRTS']
name: T-shi

In [84]:
def print_categories_with_tbo_num_for_name_last_word(tbo_num_to_print):
    set_name_last_word = set()
    
    for item_info in items_info:
        item_info['tbo'] = set()
        categories = get_categories(item_info)
        for category in categories:
            if category in dict_tbo_categories['top']:
                item_info['tbo'].add('top')
            if category in dict_tbo_categories['bottom']:
                item_info['tbo'].add('bottom')
            if category in dict_tbo_categories['onepiece']:
                item_info['tbo'].add('onepiece')
        if len(item_info['tbo']) == tbo_num_to_print:
            name_last_word = item_info['name'].split()[-1]
            set_name_last_word.add(name_last_word)
    print "set_name_last_word:", set_name_last_word 

In [86]:
print_categories_with_tbo_num_for_name_last_word(2)

set_name_last_word: set(['Dress', 'Skirt', 'Tunic', 'Inserts', 'Jumpsuit'])


In [87]:
def print_categories_with_tbo_num_for_name_first_word(tbo_num_to_print):
    set_name_first_word = set()
    
    for item_info in items_info:
        item_info['tbo'] = set()
        categories = get_categories(item_info)
        for category in categories:
            if category in dict_tbo_categories['top']:
                item_info['tbo'].add('top')
            if category in dict_tbo_categories['bottom']:
                item_info['tbo'].add('bottom')
            if category in dict_tbo_categories['onepiece']:
                item_info['tbo'].add('onepiece')
        if len(item_info['tbo']) == tbo_num_to_print:
            name_first_word = item_info['name'].split()[0]
            set_name_first_word.add(name_first_word)
    print "set_name_first_word:", set_name_first_word 

In [88]:
print_categories_with_tbo_num_for_name_first_word(2)

set_name_first_word: set(['Hooded', 'Long', 'Maxi', 'Circle', 'Cr\xc3\xaape', 'Pencil', 'Knee-length', 'Fine-knit', 'T-shirt', 'Fitted', 'V-neck', 'Striped', 'Short', 'Tunic', 'Sleeveless', 'Ribbed', 'Lyocell', 'Lyocell-blend', 'Jersey', 'Long-sleeved', 'Strapless', 'Chiffon', 'Embroidered', 'Sweatshirt', 'Jumpsuit', 'Short-sleeved', 'Cotton'])


__My conclusion__: ignore the items with more than one TBO categries.